### DISCUSSION TITLE
Hotrodding the flash

### DISCUSSION THREAD

**Discussion by myklemykle** [earlephilhower/arduino-pico#1531] [Q&A] [answered]
Hi,

My board runs a flash chip (Winbond W25Q128JW) that's not currently supported by any of the boot2 options except generic/4 .  The Winbond datasheet makes claims of great speed, of course.  What would be my starting resource for trying to puzzle out a custom set of flash timings?  I looked in the github repo, but I only see serialized pre-compiled code blocks for each of those options.

Thanks,
-mykle-

**Comment by earlephilhower** [comment, marked as answer]
The boot2 `.S` files are taken right from the Pico SDK, so I would suggest starting looking at https://github.com/raspberrypi/pico-sdk/tree/master/src/rp2_common/boot_stage2 where the original sources are from.  The .S files in here are just the assembled and padded binary versions (see `tools/libpico` for more info).

**Reply by myklemykle** [reply]
Is there a build command for compiling+converting all the .S files in boot_stage2 to the padded versions and installing them in Arduino-Pico?  Then I could just iterate in that directory and test with my current arduino-cli build setup, instead of having to get involved with the cmake system of pico-sdk .

**Reply by earlephilhower** [reply]
See https://github.com/earlephilhower/arduino-pico/blob/3950b944748950756fcf89576eefb6f08b7aea09/tools/libpico/make-libpico.sh#L25-L49

**Reply by myklemykle** [reply]
Thanks much! I have a working stage2 for my weird chip now, and performance is much better.

So far I've just been hacking this into the existing boot2 directory in the distro. I'm not sure if/when it'll ever be part of the main pico-sdk distro, so i'd like to figure out how to not have to touch the distro files, if possible. 

Do you know if there's any way to tell arduino-cli to just get my stage2 source file, or the padded+checksum version, from a path I specify? Instead of specifying it as a board option?  I'm not sure if this would be a boot property or a compiler flag or what, I just know that section of code is compiled/linked in a special specific way.

**Reply by earlephilhower** [reply]
Sorry, not that I know.  The CLI should use https://github.com/earlephilhower/arduino-pico/blob/f60b7831c88404be6927d92806c9b78706842788/platform.txt#L141-L145 so I would just make or edit a custom board and select that board using the appropriate tuple on the CLI command line.

**Comment by maxgerhardt** [comment]
For PlatformIO I also setup the repo 

https://github.com/maxgerhardt/pico-boot2-generator

This is inspired from the micropython/circuitpython repository on how `.S` files are generated using the flash information database. So if you can get your flash chip definition in there, it should spit the right `.S` file out nicely.

Note that we already have
* `Winbond/W25Q128JVxM`
* `Winbond/W25Q128JVxQ`

so yours should be pretty close to that I guess.
